### DISCUSSION TITLE
pdb loading

### DISCUSSION THREAD

**Discussion by faranak1991** [openforcefield/discussions#29] [Q&A] [answered]
Hi All,

I am new to learning about openFF, and I'm currently facing difficulties loading the pdb file (attached as pdb-file.txt) of the TBP molecule using the provided commands. Additionally, I am also attempting to load a data file from lammps instead of a pdb file. It would be greatly appreciated if someone could assist me in resolving this problem?

"
TBP = Molecule.from_smiles("O=P(OCCCC)(OCCCC)OCCCC")
topology = Topology.from_pdb( "TBP.pdb", unique_molecules=[TBP] )
"

Error:
---------------------------------------------------------------------------
UnassignedChemistryInPDBError             Traceback (most recent call last)
Cell In[149], line 2
      1 # Load the topology from a PDB file and `Molecule` objects
----> 2 topology = Topology.from_pdb( "TBP.pdb", unique_molecules=[TBP] )
      4 #topology = Topology.from_pdb("TBP.pdb")

File ~/anaconda3/envs/openff/lib/python3.11/site-packages/openff/utilities/utilities.py:80, in requires_package.<locals>.inner_decorator.<locals>.wrapper(*args, **kwargs)
     77 except Exception as e:
     78     raise e
---> 80 return function(*args, **kwargs)

File ~/anaconda3/envs/openff/lib/python3.11/site-packages/openff/toolkit/topology/topology.py:1731, in Topology.from_pdb(cls, file_path, unique_molecules, toolkit_registry, _additional_substructures)
   1725 substructure_dictionary["ADDITIONAL_SUBSTRUCTURE_OVERLAP"] = {}
   1727 coords_angstrom = np.array(
   1728     [[*vec3.value_in_unit(openmm_unit.angstrom)] for vec3 in pdb.getPositions()]
   1729 )
-> 1731 topology = toolkit_registry.call(
   1732     "_polymer_openmm_pdbfile_to_offtop",
   1733     cls,
   1734     pdb,
   1735     substructure_dictionary,
   1736     coords_angstrom,
   1737 )
   1739 for off_atom, atom in zip([*topology.atoms], pdb.topology.atoms()):
   1740     off_atom.metadata["residue_name"] = atom.residue.name

File ~/anaconda3/envs/openff/lib/python3.11/site-packages/openff/toolkit/utils/toolkit_registry.py:356, in ToolkitRegistry.call(self, method_name, raise_exception_types, *args, **kwargs)
    354             for exception_type in raise_exception_types:
    355                 if isinstance(e, exception_type):
--> 356                     raise e
    357             errors.append((toolkit, e))
    359 # No toolkit was found to provide the requested capability
    360 # TODO: Can we help developers by providing a check for typos in expected method names?

File ~/anaconda3/envs/openff/lib/python3.11/site-packages/openff/toolkit/utils/toolkit_registry.py:352, in ToolkitRegistry.call(self, method_name, raise_exception_types, *args, **kwargs)
    350 method = getattr(toolkit, method_name)
    351 try:
--> 352     return method(*args, **kwargs)
    353 except Exception as e:
    354     for exception_type in raise_exception_types:

File ~/anaconda3/envs/openff/lib/python3.11/site-packages/openff/toolkit/utils/rdkit_wrapper.py:289, in RDKitToolkitWrapper._polymer_openmm_pdbfile_to_offtop(self, topology_class, pdbfile, substructure_dictionary, coords_angstrom)
    286 from rdkit import Chem, Geometry
    288 omm_top = pdbfile.topology
--> 289 rdkit_mol = self._polymer_openmm_topology_to_rdmol(
    290     omm_top, substructure_dictionary
    291 )
    293 rdmol_conformer = Chem.Conformer()
    294 for atom_idx in range(rdkit_mol.GetNumAtoms()):

File ~/anaconda3/envs/openff/lib/python3.11/site-packages/openff/toolkit/utils/rdkit_wrapper.py:462, in RDKitToolkitWrapper._polymer_openmm_topology_to_rdmol(self, omm_top, substructure_library)
    455             symbols = sorted(
    456                 [SYMBOLS[atom.GetAtomicNum()] for atom in ref.GetAtoms()]
    457             )
    458             resname_to_symbols_and_atomnames[resname].append(
    459                 (symbols, atom_names)
    460             )
--> 462     raise UnassignedChemistryInPDBError(
    463         substructure_library=resname_to_symbols_and_atomnames,
    464         omm_top=omm_top,
    465         unassigned_atoms=unassigned_atoms,
    466         unassigned_bonds=unassigned_bonds,
    467         matches=matches,
    468     )
    470 return mol

UnassignedChemistryInPDBError: Some bonds or atoms in the input could not be identified.

Hint: The following residue names with unassigned atoms were not found in the substructure library. While the OpenFF Toolkit identifies residues by matching chemical substructures rather than by residue name, it currently only supports the 20 'canonical' amino acids.
    UNL

Error: The following 44 atoms exist in the input but could not be assigned chemical information from the substructure library:
    Atom     0 (P) in residue X:UNL#0001
    Atom     1 (O) in residue X:UNL#0001
    Atom     2 (O2) in residue X:UNL#0001
    Atom     3 (O3) in residue X:UNL#0001
    Atom     4 (O4) in residue X:UNL#0001
    Atom     5 (C) in residue X:UNL#0001
    Atom     6 (C2) in residue X:UNL#0001
    Atom     7 (C3) in residue X:UNL#0001
    Atom     8 (C4) in residue X:UNL#0001
    Atom     9 (C5) in residue X:UNL#0001
    Atom    10 (C6) in residue X:UNL#0001
    Atom    11 (C7) in residue X:UNL#0001
    Atom    12 (C8) in residue X:UNL#0001
    Atom    13 (C9) in residue X:UNL#0001
    Atom    14 (C10) in residue X:UNL#0001
    Atom    15 (C11) in residue X:UNL#0001
    Atom    16 (C12) in residue X:UNL#0001
    Atom    17 (H) in residue X:UNL#0001
    Atom    18 (H2) in residue X:UNL#0001
    Atom    19 (H3) in residue X:UNL#0001
    Atom    20 (H4) in residue X:UNL#0001
    Atom    21 (H5) in residue X:UNL#0001
    Atom    22 (H6) in residue X:UNL#0001
    Atom    23 (H7) in residue X:UNL#0001
    Atom    24 (H8) in residue X:UNL#0001
    Atom    25 (H9) in residue X:UNL#0001
    Atom    26 (H10) in residue X:UNL#0001
    Atom    27 (H11) in residue X:UNL#0001
    Atom    28 (H12) in residue X:UNL#0001
    Atom    29 (H13) in residue X:UNL#0001
    Atom    30 (H14) in residue X:UNL#0001
    Atom    31 (H15) in residue X:UNL#0001
    Atom    32 (H16) in residue X:UNL#0001
    Atom    33 (H17) in residue X:UNL#0001
    Atom    34 (H18) in residue X:UNL#0001
    Atom    35 (H19) in residue X:UNL#0001
    Atom    36 (H20) in residue X:UNL#0001
    Atom    37 (H21) in residue X:UNL#0001
    Atom    38 (H22) in residue X:UNL#0001
    Atom    39 (H23) in residue X:UNL#0001
    Atom    40 (H24) in residue X:UNL#0001
    Atom    41 (H25) in residue X:UNL#0001
    Atom    42 (H26) in residue X:UNL#0001
    Atom    43 (H27) in residue X:UNL#0001
[pdb-file.txt](https://github.com/openforcefield/discussions/files/11941049/pdb-file.txt)

**Comment by j-wags** [comment]
Hi @faranak1991,

The problem is that your PDB file is missing CONECT records indicating where the bonds should be. If possible, see if your original data source can produce a pdb file with CONECT information included. But if not, you can try letting RDKit guess where the bonds are:
```python
from rdkit import Chem
mol = Chem.MolFromPDBFile('file.pdb', removeHs=False)
Chem.MolToPDBFile(mol, 'file_CONECT.pdb')
from openff.toolkit import Molecule, Topology
TBP = Molecule.from_smiles("O=P(OCCCC)(OCCCC)OCCCC")
topology = Topology.from_pdb( "file_CONECT.pdb", unique_molecules=[TBP] )
```

This leads to a correctly-loaded molecule for me - Could you let me know if it works for you as well?

**Reply by faranak1991** [reply]
Dear Jeff

Appreciate your quick reply. It works for me as well :))

Following after conversion to LAMMPS file, shows " UnsupportedExportError: Unsupported electrostatics method Coulomb " . Is this error related to the topology?

Thanks,

**Reply by mattwthompson** [reply]
Is your topology periodic? (What is `topology.box_vectors`?) That error likely stems from gas-phase calculations not being implemented in `Interchange.to_lammps` yet.

**Reply by faranak1991** [reply]
The topology is periodicity. I appreciate you bringing up the command "topology.box_vectors," which I overlooked. By applying the following command, I found that it effectively resolved that issue.

topology.box_vectors = [30, 30, 30] * unit.angstrom
nonbondedCutoff=15* unit.angstrom

If I modify the line "pair_style lj/cut/coul/long 9.0 9.0" in the tmp.in file to " 15 A," would that be correct?

Thanks,

**Reply by mattwthompson** [reply, marked as answer]
Not necessarily - if you're using an OpenFF force field, it's been trained on a 9 A cutoff so that's the recommended use. Using a longer cutoff is sometimes desirable, but my understanding is that it's better to stick with what the force field prescribes because that probably produces more accurate physical properties, etc. Also, in case this is what you happened to be doing, you probably don't want to keep the cutoff at  half the box length, especially with larger boxes, because the cutoff error becomes vanishingly small but the extra computational cost increases significantly.

If you wanted to modify the force field and use a different cutoff, and I'm not recommending you do, you can do that by setting it on the `ForceField` object directly and earlier in your script, i.e.

```python
>>> from openff.toolkit import ForceField
>>> sage = ForceField("openff-2.1.0.offxml")
>>> sage['vdW'].cutoff
<Quantity(9.0, 'angstrom')>
>>> from openff.units import unit
>>> sage['vdW'].cutoff = 15 * unit.angstrom
>>> sage['vdW'].cutoff
<Quantity(15, 'angstrom')>
```

before you do anything like `.to_interchange`. In your example, this should produce `pair_style lj/cut/coul/long 15.0 15.0` or something similar without needing to manually edit the input script.
